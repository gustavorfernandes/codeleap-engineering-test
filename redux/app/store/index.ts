import { configureStore } from "@reduxjs/toolkit"
import usernameReducer from "../slices/usernameSlice"

export const store = configureStore({
  reducer: {
    username: usernameReducer
  },
});

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = ReturnType<typeof store.dispatch>
