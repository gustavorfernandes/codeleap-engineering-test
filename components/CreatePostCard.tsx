
import { useState } from "react"
import { useSelector, useDispatch } from "react-redux"
import createPost from "../actions/createPost"
import getPosts from "../actions/getPosts"
import { RootState } from "../redux/app/store"

function CreatePostCard() {
  const [inputTitle, setInputTitle] = useState("")
  const [inputContent, setInputContent] = useState("")
  const username = useSelector((state: RootState) => state.username.value)
  
  const dispatch = useDispatch()

  function submitForm(e: React.FormEvent) {
    e.preventDefault()
    clearFields()
  }

  function clearFields() {
    setInputTitle("")
    setInputContent("")
  }

  return (
    <div className="w-screen bg-white flex items-center justify-center py-6">
      <form
        className="w-10/12 border rounded-sm border-neutral-300 flex flex-col py-3 px-4 font-roboto"
        onSubmit={(e: React.FormEvent) => {
          submitForm(e)
        }}
      >
        <h1 className="font-bold text-sm mb-3">
          {`What's on your mind?`}
        </h1>
        <label
          className="text-xs mb-1"
          htmlFor="title"
        >
          Title
        </label>
        <input
          className="w-full text-xs outline-none border border-neutral-400 rounded py-1 px-2 mb-4"
          name="title"
          id="title"
          placeholder="Hello world"
          type="text"
          value={inputTitle}
          onChange={(e) => {            
            setInputTitle(e.target.value)            
          }}
        />
        <label
          className="text-xs mb-1"
          htmlFor="content"
        >
          Content
        </label>
        <textarea
          className="w-full text-xs outline-none border border-neutral-400 rounded py-1 px-2 mb-4 resize-none"
          name="content"
          id="content"
          rows={5}
          placeholder="Content here"
          value={inputContent}
          onChange={(e) => {
            setInputContent(`${e.target.value}`)           
          }}
        />
        <button
          className="self-end my-2 bg-black text-white font-bold text-xs px-5 py-1  disabled:bg-neutral-300"
          type="submit"
          disabled={!inputTitle || !inputContent}
          onClick={() => {            
            createPost(username, inputTitle, inputContent)
            getPosts()
          }}
        >
          CREATE
        </button>
      </form>
    </div>
  )
}

export default CreatePostCard
