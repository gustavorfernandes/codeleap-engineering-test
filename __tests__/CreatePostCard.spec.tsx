import React from "react"
import "@testing-library/jest-dom"
import { fireEvent, render, screen } from "@testing-library/react"
import { Provider } from "react-redux"
import store from "../redux/app/store"
import CreatePostCard from "../components/CreatePostCard"

describe("Create post card component", () => {
  it("When one or both inputs are empty, the create button must be disabled", () => {
    render(
      <Provider store={store} >
        <CreatePostCard />
      </Provider>
    )
    const inputTitle = screen.getByPlaceholderText("Hello world")
    const inputContent = screen.getByPlaceholderText("Content here")
    const button = screen.getByRole("button")

    expect(inputTitle).toBeInTheDocument()
    expect(inputContent).toBeInTheDocument()
    expect(button).toBeInTheDocument()
    expect(button).toBeDisabled()
        
    fireEvent.change(inputTitle, { target: { value: "" } })
    fireEvent.change(inputContent, { target: { value: "" } })
    expect(button).toBeDisabled()

    fireEvent.change(inputTitle, { target: { value: "value" } })
    fireEvent.change(inputContent, { target: { value: "" } })
    expect(button).toBeDisabled()

    fireEvent.change(inputTitle, { target: { value: "" } })
    fireEvent.change(inputContent, { target: { value: "value" } })
    expect(button).toBeDisabled()
  })

  it("When the create button is pressed, must send a post request to the API", () => {
    render(
      <Provider store={store} >
        <CreatePostCard />
      </Provider>
    )

    // Test here
  }) 

  it("The component must render correctly", () => {
    const { container } = render(
      <Provider store={store} >
        <CreatePostCard />
      </Provider>
    )
    expect(container).toMatchSnapshot()
  })
})
