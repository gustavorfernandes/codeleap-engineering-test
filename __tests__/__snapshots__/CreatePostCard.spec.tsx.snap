// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create post card component The component must render correctly 1`] = `
<div>
  <div
    class="w-screen max-w-[800px] bg-white flex items-center justify-center py-6 sm:py-10 md:py-12"
  >
    <form
      class="w-11/12 border rounded-sm border-neutral-300 flex flex-col p-4 sm:p-6 md:p-8 font-roboto"
    >
      <h1
        class="font-bold text-sm sm:text-xl md:text-[22px] mb-3 sm:mb-5"
      >
        What's on your mind?
      </h1>
      <label
        class="text-xs sm:text-sm md:text-base mb-1 sm:mb-2"
        for="title"
      >
        Title
      </label>
      <input
        class="w-full text-xs sm:text-[13px] md:text-sm outline-none border border-neutral-400 rounded py-1 px-2 mb-4"
        id="title"
        name="title"
        placeholder="Hello world"
        type="text"
        value=""
      />
      <label
        class="text-xs sm:text-sm md:text-base mb-1 sm:mb-2"
        for="content"
      >
        Content
      </label>
      <textarea
        class="w-full text-xs sm:text-[13px] md:text-sm outline-none border border-neutral-400 rounded py-1 px-2 mb-4 resize-none"
        id="content"
        name="content"
        placeholder="Content here"
        rows="4"
      />
      <button
        class="self-end mt-4 sm:mt-6 mb-2 bg-black hover:bg-neutral-800 transition-all text-white font-bold text-xs sm:text-sm md:text-base px-5 sm:px-7 py-1 disabled:bg-neutral-300"
        disabled=""
        type="submit"
      >
        CREATE
      </button>
    </form>
  </div>
</div>
`;
